### PR TITLE
IS :: STRING NOT NULL as an index compatible predicate

### PR DESCRIPTION
**Pull request by arnefischereit**

### CODE REVIEW

**Comment by neo-technology-commit-status-publisher** [comment]
Thanks for the documentation updates.

The preview documentation has now been torn down - reopening this PR will republish it.

File: modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc [hunk 1]
```diff
@@ -187,6 +187,10 @@ However, other predicates are only used when it is known that the property is co
 
 This means that a `TEXT` index is not able to solve e.g. `a.prop = b.prop`, unless a type constraint also exists on the property.
 
+One can add xref:values-and-types/type-predicate.adoc[type predicate expressions] such as `IS {two-colons} STRING NOT NULL` to enforce the String type and possibly the property's existence.
+Note that `null IS {two-colons} STRING` and that indexes do not store entities with `null` properties.
```
Comment (blouerat) [inline modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc]: Is there something missing in this sentence or is it me? I don't understand it

Comment (arnefischereit) [reply modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc]: I guess that is Mathematician slang: `null IS :: STRING` is a statement. One could add a "holds" or "is true" here, to make it less geeky:
```suggestion
Note that `null IS {two-colons} STRING` holds and that indexes do not store entities with `null` properties.
```

Comment (blouerat) [reply modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc]: Ah right, yes, I see it now! I did not read `IS` as _is_, I think it's more readable if you add "holds" or "is true" at the end.

Comment (arnefischereit) [reply modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc]: Done in 31c03ee.

**Review by blouerat — APPROVED** [review]

File: modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc [hunk 2]
```diff
@@ -187,6 +187,10 @@ However, other predicates are only used when it is known that the property is co
 
 This means that a `TEXT` index is not able to solve e.g. `a.prop = b.prop`, unless a type constraint also exists on the property.
 
+One can add xref:values-and-types/type-predicate.adoc[type predicate expressions] such as `IS {two-colons} STRING NOT NULL` to enforce the String type and possibly the property's existence.
+Note that `null IS {two-colons} STRING` is true and that indexes do not store entities with `null` properties.
+That is why, to be able to use the index, one has to use `IS {two-colons} STRING NOT NULL`.
+
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc:193]: Could the "Note that `null IS {two-colons} STRING` is true" be clarified? Particularly the "is true" part?

```suggestion
As of Neo4j 5.15, it is possible to add xref:values-and-types/type-predicate.adoc[type predicate expressions], such as `IS {two-colons} STRING NOT NULL`, to enforce the `STRING` type and the property's existence.
Note that `null IS {two-colons} STRING` is true and that indexes do not store entities with `null` properties.
That is why, to be able to use the index, `IS {two-colons} STRING NOT NULL` must be used.
```

File: modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc [hunk 3]
```diff
@@ -172,7 +172,7 @@ For more information, see xref::planning-and-tuning/query-tuning/indexes.adoc#ex
 == TEXT indexes
 
 In combination with node label and relationship type predicates, `TEXT` indexes only solve predicates operating on strings.
-That means that `TEXT` indexes are only used when it is known that the predicate evaluates to `null` for all non-string values.
+That means that `TEXT` indexes are only used when it is known that the predicate evaluates to `false` or `null` for all non-string values.
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc:175]: Going forward, we need to use this formatting when referring to Cypher types

```suggestion
That means that `TEXT` indexes are only used when it is known that the predicate evaluates to `false` or `null` for all non-`STRING` values.
```

File: modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc [hunk 4]
```diff
@@ -246,8 +258,8 @@ MATCH (n:Label) WHERE n.prop = $param
 Such queries can be modified to provide this information.
 Depending on how values that are not of type string should be treated, there are two options:
 
-* If rows in which the expression is not of type string should be discarded, then adding `WHERE <expression> STARTS WITH ''` is the right option:
-`MATCH (n:Label) WHERE $param STARTS WITH '' AND n.prop = $param`
+* If rows in which the expression is not of type string should be discarded, then adding `WHERE <expression> IS {two-colons} STRING NOT NULL` is the right option:
+`MATCH (n:Label) WHERE $param IS {two-colons} STRING NOT NULL AND n.prop = $param` (See the documentation on xref::values-and-types/type-predicate.adoc[Type predicate expressions] for more information on these predicates)
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc:262]:
```suggestion
* If it is necessary to discard rows in which the expression is not of type `STRING`, then adding `WHERE <expression> IS {two-colons} STRING NOT NULL` is the correct option. 
For example:`MATCH (n:Label) WHERE $param IS {two-colons} STRING NOT NULL AND n.prop = $param` (See the documentation on xref::values-and-types/type-predicate.adoc[Type predicate expressions] for more information on these predicates)
```

File: modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc [hunk 5]
```diff
@@ -223,11 +227,19 @@ ENDS WITH
...
 | Substring search.
 a|
-[source, syntax, role="noheader"]
+[source,syntax,role="noheader"]
 ----
 CONTAINS
 ----
 
+|
+xref:values-and-types/type-predicate.adoc[String type predicates with existence check].
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning/indexes.adoc:236]:
```suggestion
xref:values-and-types/type-predicate.adoc[`STRING` type predicates with existence check].
```

**Review by JPryce-Aklundh — COMMENTED** [review]
Thanks, @arnefischereit - I have left a few suggestions. Also, should there be an entry in the Additions and Deprecations page about this?

FYI, I am currently in the process of moving this page into a new Indexes chapter, so I will make sure to include the changes brought about by this PR in the new chapter as well.

**Review by blouerat — APPROVED** [review]
Very nice, thank you @JPryce-Aklundh

**Pull request merged**